test/core/ExamplesUnitTests/TribitsExampleApp_Tests.cmake

```diff
@@ -647,6 +647,151 @@ TribitsExampleApp_ALL_ST(ByPackage STATIC)
 TribitsExampleApp_ALL_ST(ByPackage SHARED)
 
 
+################################################################################
+
+
+function(TribitsExampleApp_INCLUDE byProjectOrPackage sharedOrStatic importedNoSystem)
+
+  if (byProjectOrPackage STREQUAL "ByProject")
+    set(findByProjectOrPackageArg -DTribitsExApp_FIND_INDIVIDUAL_PACKAGES=OFF)
+    set(foundProjectOrPackageStr "Found TribitsExProj")
+  elseif (byProjectOrPackage STREQUAL "ByPackage")
+    set(findByProjectOrPackageArg -DTribitsExApp_FIND_INDIVIDUAL_PACKAGES=ON)
+    set(foundProjectOrPackageStr "Found SimpleCxx")
+  else()
+    message(FATAL_ERROR "Invaid value for findByProjectOrPackageArg='${findByProjectOrPackageArg}'!")
+  endif()
+
+  TribitsExampleApp_process_sharedOrStatic_arg()
+
+  if (importedNoSystem STREQUAL "IMPORTED_NO_SYSTEM")
```

Comment:
>  I would use a bool instead of a string

@mperrinel, I strongly disagree.

Bools are almost always a bad choice, except for one-arg functions and some local variables where the meeting is obvious.  And bools should be avoided as arguments in multi-argument functions.  See section "12.6 Enumerated Types" in "Code Complete: 2nd Edition" which says:

> Enumerated types are especially useful for defining routine parameters

and then gives an example of a terrible function that takes multiple bools like:

```
int result = RetrievePayrollData(data, true, false, false, true)
```

There is similar advice in the book "Clean Code":

> Flag arguments are ugly.  Passing a boolean into a function is truly a terrible practices.

From the same author [Robert C. Martin's Clean Code Tip #12: Eliminate Boolean Arguments](https://www.informit.com/articles/article.aspx?p=1392524).

And then there is:

* https://alexkondov.com/should-you-pass-boolean-to-functions/
* https://understandlegacycode.com/blog/what-is-wrong-with-boolean-parameters/
* https://medium.com/swlh/solving-the-boolean-identity-crisis-33eeccde2c96

@mperrinel, where do you find guidance/best practice to have multiple bool arguments passed to a function?  What books have you read that have defined your software development practicies?

Reply:
Hi Ross,
Yes I understand your point and I read all the documentation you sent, [except the book of Robert C. Martin’s Clean Code Tip #12: Eliminate Boolean Arguments]. My point with the Boolean suggestion was in order to simplify the code for more robustness :
```
if (byProjectOrPackage STREQUAL "ByProject")
  set(findByProjectOrPackageArg -DTribitsExApp_FIND_INDIVIDUAL_PACKAGES=OFF)
  set(foundProjectOrPackageStr "Found TribitsExProj")
elseif (byProjectOrPackage STREQUAL "ByPackage")
  set(findByProjectOrPackageArg -DTribitsExApp_FIND_INDIVIDUAL_PACKAGES=ON)
  set(foundProjectOrPackageStr "Found SimpleCxx")
else()
  message(FATAL_ERROR "Invaid value for findByProjectOrPackageArg='${findByProjectOrPackageArg}'!")
endif()
```
which would have become :
```
if (isPackaged)
  set(findByProjectOrPackageArg -DTribitsExApp_FIND_INDIVIDUAL_PACKAGES=ON)
  set(foundProjectOrPackageStr "Found SimpleCxx")
else()
  set(findByProjectOrPackageArg -DTribitsExApp_FIND_INDIVIDUAL_PACKAGES=OFF)
  set(foundProjectOrPackageStr "Found TribitsExProj")
endif()
```
 In the current implementation, it is possible to make a mistake when calling the function, which will result in entering the else path. If using a Boolean instead, this mistake is no longer possible (it is a package or it is not a package).
Also, I don’t think that is it is necessarily always the best practice to hard code string values directly into a function. Indeed, if one creates another function that uses the **byProjectOrPackage** enum, one will then have to test again with the hard coded string, hereby risking to introduce typos. In addition, if one needs to change it later, that will result in more work ass all the functions using the hard coded string will have to be modified as well. Moreover, users of the function can easily introduce typos when they calls the function, which is simply impossible with the Boolean.
I understand that, with a Boolean, calling the method that way : `TribitsExampleApp_INCLUDE(TRUE SHARED IMPORTED_NO_SYSTEM)` is not good because one cannot what **TRUE** could mean in this context, and one needs to look into the function in order to understand it. However, I think that could be easily fixed by creating a variable as I suggested like that:` set(isPackage TRUE)`.
Finally, I did not realize that **byProjectOrPackage** was an enum because it seems to have only two possible values and the function uses it values to define the same CMake option **TribitsExApp_FIND_INDIVIDUAL_PACKAGES**, which contains **PACKAGES** in its name. This appeared to me as being more associated with the **ByPackage** enum value. That is why I thought that using a Boolean based on the **ByPackage** enum value named **isPackage** would be a good idea.

Reply:
> My point with the Boolean suggestion was in order to simplify the code for more robustness

So an enum has to have more than two values or you should always use a bool?  I strongly disagree with that.

Let's agree to disagree on this one

Reply:
Let's discuss this briefly at the next meeting to explain the massive advantages of passing enum-like strings in this case vs. bool arguments.

But basically, the huge advantage of using enum-like string args like this is that test definition calls like:

https://github.com/TriBITSPub/TriBITS/blob/a72c0d5266eec342751abb7b1e474706ef5b9eb2/test/core/ExamplesUnitTests/TribitsExampleApp_Tests.cmake#L390-L392

showing:

```
TribitsExampleApp_NoFortran(FULL STATIC)
TribitsExampleApp_NoFortran(COMPONENTS STATIC)
TribitsExampleApp_NoFortran(COMPONENTS SHARED)
```

exactly map to the actual CTest names like shown at [here](https://testing.sandia.gov/cdash/build/10696211/configure) showing:

```
-- TriBITS_TribitsExampleApp_NoFortran_FULL_STATIC: Added test (BASIC, PROCESSORS=1)!
-- TriBITS_TribitsExampleApp_NoFortran_COMPONENTS_STATIC: Added test (BASIC, PROCESSORS=1)!
-- TriBITS_TribitsExampleApp_NoFortran_COMPONENTS_SHARED: Added test (BASIC, PROCESSORS=1)!
```

And we see the same things with the other tests like at:

* https://github.com/TriBITSPub/TriBITS/blob/master/test/core/ExamplesUnitTests/TribitsExampleApp_Tests.cmake#L786-L790

showing:

```
TribitsExampleApp_INCLUDE(ByProject STATIC "")
TribitsExampleApp_INCLUDE(ByProject SHARED IMPORTED_NO_SYSTEM)
TribitsExampleApp_INCLUDE(ByProject STATIC IMPORTED_NO_SYSTEM)
TribitsExampleApp_INCLUDE(ByPackage SHARED IMPORTED_NO_SYSTEM)
TribitsExampleApp_INCLUDE(ByPackage STATIC IMPORTED_NO_SYSTEM)
```

and the associated tests created shown [here](https://testing.sandia.gov/cdash/build/10696211/configure) showing:

```
-- TriBITS_TribitsExampleApp_INCLUDE_ByProject_STATIC: Added test (BASIC, PROCESSORS=1)!
-- TriBITS_TribitsExampleApp_INCLUDE_ByProject_SHARED_IMPORTED_NO_SYSTEM: Added test (BASIC, PROCESSORS=1)!
-- TriBITS_TribitsExampleApp_INCLUDE_ByProject_STATIC_IMPORTED_NO_SYSTEM: Added test (BASIC, PROCESSORS=1)!
-- TriBITS_TribitsExampleApp_INCLUDE_ByPackage_SHARED_IMPORTED_NO_SYSTEM: Added test (BASIC, PROCESSORS=1)!
-- TriBITS_TribitsExampleApp_INCLUDE_ByPackage_STATIC_IMPORTED_NO_SYSTEM: Added test (BASIC, PROCESSORS=1)!
```

It is pretty much impossible to have more self-documenting test code than that.

Compare that to switching to bool args as:

```
TribitsExampleApp_INCLUDE(False False False)
TribitsExampleApp_INCLUDE(False True True)
TribitsExampleApp_INCLUDE(False False True)
TribitsExampleApp_INCLUDE(True True True)
TribitsExampleApp_INCLUDE(True False True)
```

What tests to those calls correspond to? What are they actually testing?

Can you see how terrible that is?